server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/service/ExportService.java

```diff
@@ -45,33 +48,39 @@
     private static final Logger LOGGER = LoggerFactory.getLogger(ExportService.class);
     private static final String EXPLANATION = "The backup of your mailboxes has been exported to you";
     private static final String FILE_PREFIX = "mailbox-backup-";
-    public static final boolean LAZY_RESOURCE_CLEANUP = false;
 
     private final MailboxBackup mailboxBackup;
     private final BlobStore blobStore;
     private final BlobExportMechanism blobExport;
     private final UsersRepository usersRepository;
 
-    public ExportService(MailboxBackup mailboxBackup, BlobStore blobStore, BlobExportMechanism blobExport, UsersRepository usersRepository) {
+    @Inject
+    ExportService(MailboxBackup mailboxBackup, BlobStore blobStore, BlobExportMechanism blobExport, UsersRepository usersRepository) {
         this.mailboxBackup = mailboxBackup;
         this.blobStore = blobStore;
         this.blobExport = blobExport;
         this.usersRepository = usersRepository;
     }
 
     public Mono<Task.Result> exportUserMailboxesData(Username username) {
-        return Mono.using(
-            () -> zipMailboxesContent(username),
-            inputStream -> blobStore.save(blobStore.getDefaultBucketName(), inputStream, BlobStore.StoragePolicy.LOW_COST)
+        return Mono.usingWhen(
+            Mono.fromCallable(() -> zipMailboxesContent(username)),
+            inputStream -> saveAndExportThenDeleteBlob(username, inputStream),
+            inputStream -> Mono.fromRunnable(() ->
+                Throwing.consumer(InputStream::close).sneakyThrow())
+        );
+    }
+
+    private Mono<Task.Result> saveAndExportThenDeleteBlob(Username username, InputStream inputStream) {
+        return Mono.defer(() ->
```

Comment:
Why do we have `Mono.defer` everywhere? I ask a question, I'm not sure myself, as I'm not convinced it was necessary?

Reply:
because there's some blocking code that can potentially throw an IOException in the blobStore.save() method. I need to wrap them in a Mono in order for the Exception to be handle by the onErrorResume block.